rust/lance/src/index/vector/builder.rs

```diff
@@ -412,45 +494,48 @@ impl<S: IvfSubIndex + 'static, Q: Quantization + Clone + 'static> IvfIndexBuilde
             "quantizer not set before building partition",
             location!(),
         ))?;
+        let sub_index_params = self.sub_index_params.clone().ok_or(Error::invalid_input(
+            "sub index params not set before building partition",
+            location!(),
+        ))?;
 
         // build quantized vector storage
-        let object_store = ObjectStore::local();
         let storage_len = {
             let storage = StorageBuilder::new(self.column.clone(), self.distance_type, quantizer)
                 .build(batch)?;
             let path = self.temp_dir.child(format!("storage_part{}", part_id));
-            let writer = object_store.create(&path).await?;
-            let mut writer = FileWriter::try_new(
-                writer,
-                storage.schema().as_ref().try_into()?,
-                Default::default(),
-            )?;
-            for batch in storage.to_batches()? {
-                writer.write_batch(&batch).await?;
-            }
-            writer.finish().await? as usize
+            let batches = storage.to_batches()?;
+            Self::write_batches(path, batches, storage.schema().as_ref().try_into()?).await?
         };
 
         // build the sub index, with in-memory storage
         let index_len = {
             let vectors = batch[&self.column].as_fixed_size_list();
             let flat_storage = FlatFloatStorage::new(vectors.clone(), self.distance_type);
-            let sub_index = S::index_vectors(&flat_storage, self.sub_index_params.clone())?;
+            let sub_index = S::index_vectors(&flat_storage, sub_index_params)?;
             let path = self.temp_dir.child(format!("index_part{}", part_id));
-            let writer = object_store.create(&path).await?;
             let index_batch = sub_index.to_batch()?;
-            let mut writer = FileWriter::try_new(
-                writer,
-                index_batch.schema_ref().as_ref().try_into()?,
-                Default::default(),
-            )?;
-            writer.write_batch(&index_batch).await?;
-            writer.finish().await? as usize
+            let schema = index_batch.schema().as_ref().try_into()?;
+            Self::write_batches(path, std::iter::once(index_batch), schema).await?
         };
 
         Ok((storage_len, index_len))
     }
 
+    async fn write_batches(
+        path: Path,
+        batches: impl Iterator<Item = RecordBatch>,
+        schema: Schema,
+    ) -> Result<usize> {
+        let object_store = ObjectStore::local();
+        let writer = object_store.create(&path).await?;
+        let mut writer = FileWriter::try_new(writer, schema, Default::default())?;
+        for batch in batches {
+            writer.write_batch(&batch).await?;
+        }
+        Ok(writer.finish().await? as usize)
+    }
```

Comment:
Doesn't have to be part of this PR but it might be nice to have this as a static method on `FileWriter`.